File: src/install/resolvers/folder_resolver.zig
const std = @import("std");
const PackageID = @import("../install.zig").PackageID;
const Lockfile = @import("../install.zig").Lockfile;
const initializeStore = @import("../install.zig").initializeStore;
const json_parser = bun.JSON;
const PackageManager = @import("../install.zig").PackageManager;
const Npm = @import("../npm.zig");
const logger = bun.logger;
const FileSystem = @import("../../fs.zig").FileSystem;
const JSAst = bun.JSAst;
const string = bun.string;
const stringZ = bun.stringZ;
const Features = @import("../install.zig").Features;
const IdentityContext = @import("../../identity_context.zig").IdentityContext;
const strings = bun.strings;
const Resolution = @import("../resolution.zig").Resolution;
const String = bun.Semver.String;
const Semver = bun.Semver;
const bun = @import("root").bun;
const Dependency = @import("../dependency.zig");
pub const FolderResolution = union(Tag) {
    package_id: PackageID,
    err: anyerror,
    new_package_id: PackageID,

    pub const Tag = enum { package_id, err, new_package_id };

    pub const PackageWorkspaceSearchPathFormatter = struct {
        manager: *PackageManager,
        version: Dependency.Version,
        quoted: bool = true,

        pub fn format(this: PackageWorkspaceSearchPathFormatter, comptime fmt: []const u8, opts: std.fmt.FormatOptions, writer: anytype) !void {
            var joined: [bun.MAX_PATH_BYTES + 2]u8 = undefined;
            const str_to_use = this.manager.lockfile.workspace_paths.getPtr(
                @truncate(String.Builder.stringHash(this.manager.lockfile.str(&this.version.value.workspace))),
            ) orelse &this.version.value.workspace;
            var paths = normalizePackageJSONPath(.{ .relative = .workspace }, joined[2..], this.manager.lockfile.str(str_to_use));

            if (!strings.startsWithChar(paths.rel, '.') and !strings.startsWithChar(paths.rel, std.fs.path.sep)) {
                joined[0..2].* = ("." ++ std.fs.path.sep_str).*;
                paths.rel = joined[0 .. paths.rel.len + 2];
            }

            if (this.quoted) {
                const quoted = bun.fmt.QuotedFormatter{
                    .text = paths.rel,
                };
                try quoted.format(fmt, opts, writer);
            } else {
                try writer.writeAll(paths.rel);
            }
        }
    };

    pub const Map = std.HashMapUnmanaged(u64, FolderResolution, IdentityContext(u64), 80);

    pub fn normalize(path: string) string {
        return FileSystem.instance.normalize(path);
    }

    pub fn hash(normalized_path: string) u64 {
        return bun.hash(normalized_path);
    }

    fn NewResolver(comptime tag: Resolution.Tag) type {
        return struct {
            folder_path: string,

            pub fn resolve(this: @This(), comptime Builder: type, builder: Builder, _: JSAst.Expr) !Resolution {
                return Resolution{
                    .tag = tag,
                    .value = @unionInit(Resolution.Value, @tagName(tag), builder.append(String, this.folder_path)),
                };
            }

            pub fn count(this: @This(), comptime Builder: type, builder: Builder, _: JSAst.Expr) void {
                builder.count(this.folder_path);
            }

            pub fn checkBundledDependencies() bool {
                return tag == .folder or tag == .symlink;
            }
        };
    }

    const Resolver = NewResolver(Resolution.Tag.folder);
    const SymlinkResolver = NewResolver(Resolution.Tag.symlink);
    const WorkspaceResolver = NewResolver(Resolution.Tag.workspace);
    const CacheFolderResolver = struct {
        version: Semver.Version,

        pub fn resolve(this: @This(), comptime Builder: type, _: Builder, _: JSAst.Expr) !Resolution {
            return Resolution{
                .tag = Resolution.Tag.npm,
                .value = .{
                    .npm = .{
                        .version = this.version,
                        .url = String.from(""),
                    },
                },
            };
        }

        pub fn count(_: @This(), comptime Builder: type, _: Builder, _: JSAst.Expr) void {}

        pub fn checkBundledDependencies() bool {
            return true;
        }
    };

    const Paths = struct {
        abs: stringZ,
        rel: string,
    };
    fn normalizePackageJSONPath(global_or_relative: GlobalOrRelative, joined: *bun.PathBuffer, non_normalized_path: string) Paths {
        var abs: string = "";
        var rel: string = "";
        // We consider it valid if there is a package.json in the folder
        const normalized = if (non_normalized_path.len == 1 and non_normalized_path[0] == '.')
            non_normalized_path
        else if (std.fs.path.isAbsolute(non_normalized_path))
            std.mem.trimRight(u8, non_normalized_path, std.fs.path.sep_str)
        else
            std.mem.trimRight(u8, normalize(non_normalized_path), std.fs.path.sep_str);

        if (strings.startsWithChar(normalized, '.')) {
            var tempcat: bun.PathBuffer = undefined;

            bun.copy(u8, &tempcat, normalized);
            tempcat[normalized.len..][0.."/package.json".len].* = (std.fs.path.sep_str ++ "package.json").*;
            var parts = [_]string{ FileSystem.instance.top_level_dir, tempcat[0 .. normalized.len + "/package.json".len] };
            abs = FileSystem.instance.absBuf(&parts, joined);
            rel = FileSystem.instance.relative(FileSystem.instance.top_level_dir, abs[0 .. abs.len - "/package.json".len]);
        } else {
            var remain: []u8 = joined[0..];
            switch (global_or_relative) {
                .global, .cache_folder => {
                    const path = if (global_or_relative == .global) global_or_relative.global else global_or_relative.cache_folder;
                    if (path.len > 0) {
                        const offset = path.len -| @as(usize, @intFromBool(path[path.len -| 1] == std.fs.path.sep));
                        if (offset > 0)
                            @memcpy(remain[0..offset], path[0..offset]);
                        remain = remain[offset..];
                        if (normalized.len > 0) {
                            if ((path[path.len - 1] != std.fs.path.sep) and (normalized[0] != std.fs.path.sep)) {
                                remain[0] = std.fs.path.sep;
                                remain = remain[1..];
                            }
                        }
                    }
                },
                .relative => {},
            }
            bun.copy(u8, remain, normalized);
            remain[normalized.len..][0.."/package.json".len].* = (std.fs.path.sep_str ++ "package.json").*;
            remain = remain[normalized.len + "/package.json".len ..];
            abs = joined[0 .. joined.len - remain.len];
            // We store the folder name without package.json
            rel = FileSystem.instance.relative(FileSystem.instance.top_level_dir, abs[0 .. abs.len - "/package.json".len]);
        }
        joined[abs.len] = 0;

        return .{
            .abs = joined[0..abs.len :0],
            .rel = rel,
        };
    }

    fn readPackageJSONFromDisk(
        manager: *PackageManager,
        abs: stringZ,
        version: Dependency.Version,
        comptime features: Features,
        comptime ResolverType: type,
        resolver: *ResolverType,
    ) !Lockfile.Package {
        var body = Npm.Registry.BodyPool.get(manager.allocator);
        defer Npm.Registry.BodyPool.release(body);

        var package = Lockfile.Package{};

        if (comptime ResolverType == WorkspaceResolver) {
            const tracer = bun.perf.trace("FolderResolver.readPackageJSONFromDisk.workspace");
            defer tracer.end();

            const json = try manager.workspace_package_json_cache.getWithPath(manager.allocator, manager.log, abs, .{}).unwrap();

            try package.parseWithJSON(
                manager.lockfile,
                manager,
                manager.allocator,
                manager.log,
                json.source,
                json.root,
                ResolverType,
                resolver,
                features,
            );
        } else {
            const tracer = bun.perf.trace("FolderResolver.readPackageJSONFromDisk.folder");
            defer tracer.end();

            const source = brk: {
                var file = bun.sys.File.from(try bun.sys.openatA(
                    bun.FD.cwd(),
                    abs,
                    bun.O.RDONLY,
                    0,
                ).unwrap());
                defer file.close();

                {
                    body.data.reset();
                    var man = body.data.list.toManaged(manager.allocator);
                    defer body.data.list = man.moveToUnmanaged();
                    _ = try file.readToEndWithArrayList(&man, true).unwrap();
                }

                break :brk logger.Source.initPathString(abs, body.data.list.items);
            };

            try package.parse(
                manager.lockfile,
                manager,
                manager.allocator,
                manager.log,
                source,
                ResolverType,
                resolver,
                features,
            );
        }

        const has_scripts = package.scripts.hasAny() or brk: {
            const dir = std.fs.path.dirname(abs) orelse "";
            const binding_dot_gyp_path = bun.path.joinAbsStringZ(
                dir,
                &[_]string{"binding.gyp"},
                .auto,
            );
            break :brk bun.sys.exists(binding_dot_gyp_path);
        };

        package.meta.setHasInstallScript(has_scripts);

        if (manager.lockfile.getPackageID(package.name_hash, version, &package.resolution)) |existing_id| {
            package.meta.id = existing_id;
            manager.lockfile.packages.set(existing_id, package);
            return manager.lockfile.packages.get(existing_id);
        }

        return manager.lockfile.appendPackage(package);
    }

    pub const GlobalOrRelative = union(enum) {
        global: []const u8,
        relative: Dependency.Version.Tag,
        cache_folder: []const u8,
    };

    pub fn getOrPut(global_or_relative: GlobalOrRelative, version: Dependency.Version, non_normalized_path: string, manager: *PackageManager) FolderResolution {
        var joined: bun.PathBuffer = undefined;
        const paths = normalizePackageJSONPath(global_or_relative, &joined, non_normalized_path);
        const abs = paths.abs;
        const rel = paths.rel;

        // replace before getting hash. rel may or may not be contained in abs
        if (comptime bun.Environment.isWindows) {
            bun.path.dangerouslyConvertPathToPosixInPlace(u8, @constCast(abs));
            bun.path.dangerouslyConvertPathToPosixInPlace(u8, @constCast(rel));
        }
        const abs_hash = hash(abs);

        const entry = manager.folders.getOrPut(manager.allocator, abs_hash) catch unreachable;
        if (entry.found_existing) return entry.value_ptr.*;

        const package: Lockfile.Package = switch (global_or_relative) {
            .global => global: {
                var path: bun.PathBuffer = undefined;
                std.mem.copyForwards(u8, &path, non_normalized_path);
                var resolver: SymlinkResolver = .{
                    .folder_path = path[0..non_normalized_path.len],
                };
                break :global readPackageJSONFromDisk(
                    manager,
                    abs,
                    version,
                    Features.link,
                    SymlinkResolver,
                    &resolver,
                );
            },
            .relative => |tag| switch (tag) {
                .folder => folder: {
                    var resolver: Resolver = .{
                        .folder_path = rel,
                    };
                    break :folder readPackageJSONFromDisk(
                        manager,
                        abs,
                        version,
                        Features.folder,
                        Resolver,
                        &resolver,
                    );
                },
                .workspace => workspace: {
                    var resolver: WorkspaceResolver = .{
                        .folder_path = rel,
                    };
                    break :workspace readPackageJSONFromDisk(
                        manager,
                        abs,
                        version,
                        Features.workspace,
                        WorkspaceResolver,
                        &resolver,
                    );
                },
                else => unreachable,
            },
            .cache_folder => cache_folder: {
                var resolver: CacheFolderResolver = .{
                    .version = version.value.npm.version.toVersion(),
                };
                break :cache_folder readPackageJSONFromDisk(
                    manager,
                    abs,
                    version,
                    Features.npm,
                    CacheFolderResolver,
                    &resolver,
                );
            },
        } catch |err| {
            if (err == error.FileNotFound or err == error.ENOENT) {
                entry.value_ptr.* = .{ .err = error.MissingPackageJSON };
            } else {
                entry.value_ptr.* = .{ .err = err };
            }

            return entry.value_ptr.*;
        };

        entry.value_ptr.* = .{ .package_id = package.meta.id };
        return FolderResolution{ .new_package_id = package.meta.id };
    }
};

File: src/install/resolvers/folder_resolver.rs
use core::fmt;

use bun_core::fmt::QuotedFormatter;
use bun_core::{ZStr, strings};
use bun_paths::{self, PathBuffer, SEP, SEP_STR, resolve_path};
use bun_resolver::fs::FileSystem;
use bun_semver::{self as semver, String as SemverString};
use bun_sys::{self, Fd, File, O};

use crate::bun_json::Expr;
use crate::dependency::{self};
use crate::install::{Features, Lockfile, PackageID};
use crate::lockfile::Package as LockfilePackage;
use crate::lockfile_real::StringBuilder;
use crate::lockfile_real::package::ResolverContext;
use crate::npm;
use crate::package_manager_real::PackageManager;
use crate::resolution::{ResolutionType, Tag as ResolutionTag, TaggedValue};
use crate::versioned_url::VersionedURLType;

#[derive(Copy, Clone)]
pub enum FolderResolution {
    PackageId(PackageID),
    Err(crate::Error),
    NewPackageId(PackageID),
}

// The enum discriminant serves as the tag; expose an alias for it.

pub(crate) struct PackageWorkspaceSearchPathFormatter<'a> {
    pub manager: &'a PackageManager,
    pub version: dependency::Version,
    pub quoted: bool,
}

impl<'a> fmt::Display for PackageWorkspaceSearchPathFormatter<'a> {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        // Caller constructs this formatter only when
        // `self.version.tag == .workspace`.
        let workspace = self.version.workspace();
        let str_to_use = self
            .manager
            .lockfile
            .workspace_paths
            .get(&semver::string::Builder::string_hash(
                self.manager.lockfile.str(workspace),
            ))
            .unwrap_or(workspace);

        let search_path = self.manager.lockfile.str(str_to_use);

        let mut joined = PathBuffer::uninit();
        let mut dot_slash_rel = Vec::new();
        let rel: &[u8] = match normalize_package_json_path(
            GlobalOrRelative::Relative(dependency::version::Tag::Workspace),
            &mut joined,
            search_path,
        ) {
            Some(paths)
                if !strings::starts_with_char(paths.rel, b'.')
                    && !strings::starts_with_char(paths.rel, SEP) =>
            {
                dot_slash_rel.push(b'.');
                dot_slash_rel.push(SEP);
                dot_slash_rel.extend_from_slice(paths.rel);
                dot_slash_rel.as_slice()
            }
            Some(paths) => paths.rel,
            // Too long to be a path; show it as written.
            None => search_path,
        };

        if self.quoted {
            let quoted = QuotedFormatter { text: rel };
            fmt::Display::fmt(&quoted, f)
        } else {
            // `fmt::Formatter` only accepts `&str`, so non-UTF-8 path bytes are emitted lossily
            // (U+FFFD) via `bstr::BStr`'s Display. Both current callers pass
            // `quoted = true`, so this branch is unreached today; if a future
            // caller needs byte-exact output it must use an `io::Write` sink.
            write!(f, "{}", bstr::BStr::new(rel))
        }
    }
}

/// Value stored in the folder-resolution map: the resolution plus the
/// normalized absolute `package.json` path the key hash was computed from.
/// Lookups compare the path, since a different path whose hash collides must
/// not reuse this resolution.
pub struct Entry {
    pub(crate) abs_path: Box<[u8]>,
    pub(crate) resolution: FolderResolution,
}

// bun_collections::HashMap currently ignores the context/load-factor
// type params (backed by std HashMap); identity hashing is a TODO(perf).

pub(crate) fn hash(normalized_path: &[u8]) -> u64 {
    bun_wyhash::hash(normalized_path)
}

// ── NewResolver ───────────────────────────────────────────────────────────
// The const-generic tag requires `#[derive(ConstParamTy)]` (already on `Tag`).
struct NewResolver<'a, const TAG: ResolutionTag> {
    pub(crate) folder_path: &'a [u8],
}

impl<'a, const TAG: ResolutionTag> ResolverContext for NewResolver<'a, TAG> {
    fn check_bundled_dependencies() -> bool {
        matches!(TAG, ResolutionTag::Folder | ResolutionTag::Symlink)
    }

    fn count(&mut self, builder: &mut StringBuilder<'_>, _json: &Expr) {
        builder.count(self.folder_path);
    }

    fn resolve(
        &mut self,
        builder: &mut StringBuilder<'_>,
        _json: &Expr,
    ) -> crate::Result<ResolutionType<u64>> {
        let appended = builder.append::<SemverString>(self.folder_path);
        Ok(ResolutionType::<u64>::init(match TAG {
            ResolutionTag::Folder => TaggedValue::Folder(appended),
            ResolutionTag::Symlink => TaggedValue::Symlink(appended),
            ResolutionTag::Workspace => TaggedValue::Workspace(appended),
            _ => unreachable!(),
        }))
    }
}

type Resolver<'a> = NewResolver<'a, { ResolutionTag::Folder }>;
type SymlinkResolver<'a> = NewResolver<'a, { ResolutionTag::Symlink }>;
type WorkspaceResolver<'a> = NewResolver<'a, { ResolutionTag::Workspace }>;

pub(crate) struct CacheFolderResolver {
    pub version: semver::Version,
}

impl ResolverContext for CacheFolderResolver {
    fn check_bundled_dependencies() -> bool {
        true
    }

    fn count(&mut self, _builder: &mut StringBuilder<'_>, _json: &Expr) {}

    fn resolve(
        &mut self,
        _builder: &mut StringBuilder<'_>,
        _json: &Expr,
    ) -> crate::Result<ResolutionType<u64>> {
        Ok(ResolutionType::<u64>::init(TaggedValue::Npm(
            VersionedURLType {
                version: self.version,
                url: SemverString::from(b""),
            },
        )))
    }
}

/// Unifies `NewResolver<TAG>` and `CacheFolderResolver` for
/// `read_package_json_from_disk`; the associated const `IS_WORKSPACE`
/// distinguishes the workspace resolver.
trait FolderResolverImpl: ResolverContext {
    const IS_WORKSPACE: bool;
}
impl<'a, const TAG: ResolutionTag> FolderResolverImpl for NewResolver<'a, TAG> {
    const IS_WORKSPACE: bool = matches!(TAG, ResolutionTag::Workspace);
}
impl FolderResolverImpl for CacheFolderResolver {
    const IS_WORKSPACE: bool = false;
}

struct Paths<'a> {
    abs: &'a ZStr,
    rel: &'a [u8],
}

/// Returns `None` when the `package.json` path does not fit `joined`.
fn normalize_package_json_path<'a>(
    global_or_relative: GlobalOrRelative<'_>,
    joined: &'a mut PathBuffer,
    non_normalized_path: &[u8],
) -> Option<Paths<'a>> {
    let mut normalize_spill = Vec::new();
    // We consider it valid if there is a package.json in the folder
    let normalized: &[u8] = if non_normalized_path == b"." {
        non_normalized_path
    } else if bun_paths::is_absolute(non_normalized_path) {
        strings::trim_right(non_normalized_path, SEP_STR.as_bytes())
    } else {
        strings::trim_right(
            resolve_path::normalize_string_spill::<true, bun_paths::platform::Auto>(
                &mut normalize_spill,
                non_normalized_path,
            ),
            SEP_STR.as_bytes(),
        )
    };

    const PACKAGE_JSON_LEN: usize = "/package.json".len();

    // The last byte of `joined` is reserved for the NUL terminator.
    let capacity = joined.len() - 1;

    let abs_len = if strings::starts_with_char(normalized, b'.') {
        let parts: [&[u8]; 2] = [normalized, b"package.json"];
        FileSystem::instance()
            .abs_buf_checked(&parts, &mut joined[..capacity])?
            .len()
    } else {
        let (prefix, needs_sep): (&[u8], bool) = match global_or_relative {
            GlobalOrRelative::Global(path) | GlobalOrRelative::CacheFolder(path)
                if !path.is_empty() =>
            {
                let ends_with_sep = path[path.len() - 1] == SEP;
                (
                    &path[..path.len() - ends_with_sep as usize],
                    !normalized.is_empty() && !ends_with_sep && normalized[0] != SEP,
                )
            }
            _ => (b"", false),
        };
        let abs_len = prefix.len() + needs_sep as usize + normalized.len() + PACKAGE_JSON_LEN;
        if abs_len > capacity {
            return None;
        }

        let mut len = prefix.len();
        joined[..len].copy_from_slice(prefix);
        if needs_sep {
            joined[len] = SEP;
            len += 1;
        }
        joined[len..len + normalized.len()].copy_from_slice(normalized);
        len += normalized.len();
        joined[len] = SEP;
        joined[len + 1..len + PACKAGE_JSON_LEN].copy_from_slice(b"package.json");
        abs_len
    };

    // We store the folder name without package.json
    let rel = FileSystem::instance().relative(
        FileSystem::instance().top_level_dir(),
        &joined[..abs_len - PACKAGE_JSON_LEN],
    );
    joined[abs_len] = 0;

    Some(Paths {
        abs: ZStr::from_buf(joined, abs_len),
        rel,
    })
}

fn read_package_json_from_disk<R: FolderResolverImpl>(
    manager: &mut PackageManager,
    abs: &ZStr,
    version: &dependency::Version,
    features: Features,
    resolver: &mut R,
) -> crate::Result<LockfilePackage> {
    let mut body = npm::Registry::BodyPool::get();

    let mut package: LockfilePackage = Default::default();

    // Borrow splitting: `manager.lockfile`, `manager`, and `manager.log` are
    // needed simultaneously; borrowck rejects the overlap on `&mut self`,
    // so split via raw pointer once here. `lockfile` and `log` are disjoint
    // fields of `PackageManager`, and `parse{,_with_json}` only reaches
    // `manager` through the `pm` argument (no re-entrant access to
    // `lockfile`/`log` via `pm`).
    //
    // `log_mut()` reads the BACKREF `self.log: *mut Log` and returns the
    // disjoint CLI `Log` allocation (lifetime decoupled from `&self`); call it
    // safely *before* establishing `manager_ptr` so `log` is derived from a
    // separate allocation and is unaffected by the `&mut *manager_ptr`
    // reborrows below.
    let log: &mut bun_ast::Log = manager.log_mut();
    let manager_ptr: *mut PackageManager = manager;

    if R::IS_WORKSPACE {
        let _tracer =
            bun_perf::trace(bun_perf::PerfEvent::FolderResolverReadPackageJSONFromDiskWorkspace);

        // SAFETY: `manager_ptr` was just derived from the live `&mut PackageManager`
        // argument; `log` points into a separate `Log` allocation (see the
        // borrow-splitting comment above), so this `&mut` reborrow aliases no
        // other live reference.
        let json = unsafe { &mut *manager_ptr }
            .workspace_package_json_cache
            .get_with_path(log, abs.as_bytes(), Default::default())
            .unwrap()?;
        // `Expr` is `Copy`; take a raw pointer to `source` so the borrow on
        // `workspace_package_json_cache` ends before `&mut *manager_ptr` is
        // formed for `parse_with_json`.
        let root: Expr = json.root;
        let source: *const bun_ast::Source = &raw const json.source;

        // SAFETY: see the borrow-splitting comment above.
        unsafe {
            let lockfile: *mut Lockfile = &raw mut *(*manager_ptr).lockfile;
            package.parse_with_json::<R>(
                &mut *lockfile,
                &mut *manager_ptr,
                log,
                &*source,
                root,
                resolver,
                features,
            )?;
        }
    } else {
        let _tracer =
            bun_perf::trace(bun_perf::PerfEvent::FolderResolverReadPackageJSONFromDiskFolder);

        let source = {
            let file = File::openat(Fd::cwd(), abs.as_bytes(), O::RDONLY, 0)?;
            body.reset();
            let read_result = file
                .read_to_end_with_array_list(&mut body.list, bun_sys::SizeHint::ProbablySmall)
                .map(|_| ());
            let _ = file.close();
            read_result?;

            bun_ast::Source::init_path_string(abs.as_bytes(), body.list.as_slice())
        };

        // SAFETY: see the borrow-splitting comment above.
        unsafe {
            let lockfile: *mut Lockfile = &raw mut *(*manager_ptr).lockfile;
            package.parse::<R>(
                &mut *lockfile,
                &mut *manager_ptr,
                log,
                &source,
                resolver,
                features,
            )?;
        }
    }

    let has_scripts = package.scripts.has_any()
        || 'brk: {
            let dir = bun_paths::dirname(abs.as_bytes()).unwrap_or(b"");
            let binding_dot_gyp_path = bun_paths::resolve_path::join_abs_string_z::<
                bun_paths::platform::Auto,
            >(dir, &[b"binding.gyp" as &[u8]]);
            break 'brk bun_sys::exists(binding_dot_gyp_path.as_bytes());
        };

    package.meta.set_has_install_script(has_scripts);

    if let Some(existing_id) =
        manager
            .lockfile
            .get_package_id(package.name_hash, Some(version), &package.resolution)
    {
        package.meta.id = existing_id;
        manager.lockfile.packages.set(existing_id as usize, package);
        return Ok(*manager.lockfile.packages.get(existing_id as usize));
    }

    Ok(manager.lockfile.append_package(&package)?)
}

#[derive(Copy, Clone)]
pub enum GlobalOrRelative<'a> {
    Global(&'a [u8]),
    Relative(dependency::version::Tag),
    CacheFolder(&'a [u8]),
}

pub(crate) fn get_or_put(
    global_or_relative: GlobalOrRelative<'_>,
    version: &dependency::Version,
    non_normalized_path: &[u8],
    manager: &mut PackageManager,
) -> FolderResolution {
    let mut joined = PathBuffer::uninit();
    #[cfg(windows)]
    let mut rel_buf = PathBuffer::uninit();
    let Some(paths) =
        normalize_package_json_path(global_or_relative, &mut joined, non_normalized_path)
    else {
        return FolderResolution::Err(crate::Error::Sys(bun_errno::SystemErrno::ENAMETOOLONG));
    };

    #[cfg(not(windows))]
    let abs = paths.abs;
    #[cfg(not(windows))]
    let rel = paths.rel;

    // replace before getting hash. rel may or may not be contained in abs
    #[cfg(windows)]
    let (abs, rel): (&ZStr, &[u8]) = {
        // Writing in place through `(&ZStr).as_ptr().cast_mut()` /
        // `(&[u8]).as_ptr().cast_mut()` would be UB under Stacked/Tree
        // Borrows: those pointers carry read-only provenance, and the
        // optimizer may assume `abs`'s bytes are unchanged when computing
        // `hash(abs.as_bytes())` below.
        //
        // Instead: capture lengths, let the shared borrows of `joined` die,
        // then take a fresh `&mut joined[..abs_len]` (write provenance) and
        // mutate that. `rel` points into FileSystem's thread-local relative
        // buffer which we only ever see as `&[u8]`, so copy it into a local
        // we own and convert the copy — same pattern as
        // WorkspacePackageJSONCache::get_with_path.
        let abs_len = paths.abs.len();
        let rel_len = paths.rel.len();
        rel_buf[..rel_len].copy_from_slice(paths.rel);
        // `paths` is dead past this point → `joined` is no longer borrowed.
        bun_paths::dangerously_convert_path_to_posix_in_place::<u8>(&mut joined[..abs_len]);
        bun_paths::dangerously_convert_path_to_posix_in_place::<u8>(&mut rel_buf[..rel_len]);
        (
            // `normalize_package_json_path` wrote `joined[abs_len] = 0`; the
            // separator rewrite above never touches the NUL.
            ZStr::from_buf(&joined[..], abs_len),
            &rel_buf[..rel_len],
        )
    };
    let abs_hash = hash(abs.as_bytes());

    // Check first, compute, then insert, because read_package_json_from_disk
    // needs &mut manager. Compare the stored path, not just its hash: a
    // different path whose hash collides must not reuse this resolution. On a
    // collision, resolve fresh without caching so the first path's entry stays.
    let hash_collision = match manager.folders.get(&abs_hash) {
        Some(existing) if *existing.abs_path == *abs.as_bytes() => return existing.resolution,
        Some(_) => true,
        None => false,
    };

    let result: crate::Result<LockfilePackage> = match global_or_relative {
        GlobalOrRelative::Global(_) => 'global: {
            // `non_normalized_path` may alias the lockfile string buffer, which grows below.
            let folder_path: Box<[u8]> = Box::from(non_normalized_path);
            let mut resolver: SymlinkResolver = NewResolver {
                folder_path: &folder_path,
            };
            break 'global read_package_json_from_disk(
                manager,
                abs,
                version,
                Features::LINK,
                &mut resolver,
            );
        }
        GlobalOrRelative::Relative(tag) => match tag {
            dependency::version::Tag::Folder => 'folder: {
                let mut resolver: Resolver = NewResolver { folder_path: rel };
                break 'folder read_package_json_from_disk(
                    manager,
                    abs,
                    version,
                    Features::FOLDER,
                    &mut resolver,
                );
            }
            dependency::version::Tag::Workspace => 'workspace: {
                let mut resolver: WorkspaceResolver = NewResolver { folder_path: rel };
                break 'workspace read_package_json_from_disk(
                    manager,
                    abs,
                    version,
                    Features::WORKSPACE,
                    &mut resolver,
                );
            }
            _ => unreachable!(),
        },
        GlobalOrRelative::CacheFolder(_) => 'cache_folder: {
            let mut resolver = CacheFolderResolver {
                // `GlobalOrRelative::CacheFolder` is only passed by
                // `PackageManagerResolution` with a `version.tag == .npm`
                // dependency.
                version: version.npm().version.to_version(),
            };
            break 'cache_folder read_package_json_from_disk(
                manager,
                abs,
                version,
                Features::NPM,
                &mut resolver,
            );
        }
    };

    let package = match result {
        Ok(p) => p,
        Err(err) => {
            let stored = if err == crate::Error::Sys(bun_errno::SystemErrno::ENOENT) {
                FolderResolution::Err(crate::Error::MissingPackageJSON)
            } else {
                FolderResolution::Err(err)
            };
            if !hash_collision {
                manager.folders.insert(
                    abs_hash,
                    Entry {
                        abs_path: abs.as_bytes().into(),
                        resolution: stored,
                    },
                );
            }
            return stored;
        }
    };

    if !hash_collision {
        manager.folders.insert(
            abs_hash,
            Entry {
                abs_path: abs.as_bytes().into(),
                resolution: FolderResolution::PackageId(package.meta.id),
            },
        );
    }
    FolderResolution::NewPackageId(package.meta.id)
}
